src/experimental/tools/ContentDownloader/api/downloader/segment.ts

```diff
@@ -0,0 +1,350 @@
+/**
+ * Copyright 2019 CANAL+ Group
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { EMPTY, merge, Observable, of, Subject } from "rxjs";
+import { distinct, map, mergeMap, reduce, scan, takeUntil, tap } from "rxjs/operators";
+
+import { ISegmentParserResponse } from "../../../../../transports";
+import find from "../../../../../utils/array_find";
+import findIndex from "../../../../../utils/array_find_index";
+import { concat, strToBytes } from "../../../../../utils/byte_parsing";
+
+import { createBox } from "../../../../../parsers/containers/isobmff";
+import { IndexedDBError } from "../../utils";
+import { ContentType } from "../tracksPicker/types";
+import {
+  ContentBufferType,
+  IAbstractContextCreation,
+  IContext,
+  IContextRicher,
+  ICustomSegment,
+  IGlobalContext,
+  IInitGroupedSegments,
+  IManifestDBState,
+  ISegmentData,
+  ISegmentPipelineContext,
+  IUtils
+} from "./types";
+
+/**
+ * Download a segment associated with a context.
+ *
+ * @remarks
+ * We are downloading the segment 3 by 3 for now.
+ * It's something that will look soon.
+ *
+ * @param {IContext[]} ctxs An array of segments context to download.
+ * > It's possibly a number, when the segment context has already been download
+ * @param {KeyContextType} contentType Tell what type of buffer it is (VIDEO/AUDIO/TEXT).
+ * @param {ISegmentPipelineContext} - Segment pipeline context, values that are redundant.
+ * @returns {Observable<ICustomSegment>} - An observable of a downloaded segments context.
+ *
+ */
+export function handleSegmentPipelineFromContexts<
+  KeyContextType extends keyof Omit<IGlobalContext, "manifest">
+>(
+  ctxs: IContext[],
+  contentType: KeyContextType,
+  {
+    segmentPipelineCreator,
+    isInitData,
+    nextSegments,
+    progress,
+    type,
+  }: ISegmentPipelineContext
+): Observable<ICustomSegment> {
+  const segmentFetcherForCurrentContentType = segmentPipelineCreator.createSegmentFetcher(
+    contentType,
+    new Subject()
+  );
+  return of(...ctxs).pipe(
+    mergeMap(
+      (ctx, index) => {
+        if (Array.isArray(ctx.segment)) {
+          return EMPTY;
+        }
+        return segmentFetcherForCurrentContentType.createRequest(ctx, 0).pipe(
+          mergeMap(evt => {
+            switch (evt.type) {
+              case "chunk":
+                return evt.parse();
+              default:
```

Comment:
We technically might want to reset the chunks accumulated in the case of an "interrupted" event no?

They are not supposed to happen here, but they technically could in the future